security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -267,6 +287,39 @@ func (s *sdsservice) FetchSecrets(ctx context.Context, discReq *xdsapi.Discovery
 	return sdsDiscoveryResponse(secret, discReq.Node.Id)
 }
 
+func (s *sdsservice) Stop() {
+	s.closing <- true
+}
+
+func (s *sdsservice) clearStaledClientsJob() {
+	s.ticker = time.NewTicker(s.tickerInterval)
+	for {
+		select {
+		case <-s.ticker.C:
+			s.clearStaledClients()
+		case <-s.closing:
+			if s.ticker != nil {
+				s.ticker.Stop()
+			}
+		}
+	}
+}
+
+func (s *sdsservice) clearStaledClients() {
+	log.Debug("start staled connection cleanup job")
+	sdsClientsMutex.Lock()
+	defer sdsClientsMutex.Unlock()
+
+	sLen := len(staledClientKeys)
+	for i := 0; i < sLen; i++ {
+		log.Debugf("remove staled clients %+v", staledClientKeys[i])
+		delete(sdsClients, staledClientKeys[i])
+	}
+
+	// truncate staledClientKeys to remove keys that deleted above.
+	staledClientKeys = staledClientKeys[sLen:]
```

Comment:
Not sure what this does, since sLen equals to len(staledClientKeys)

Reply:
before delete we take a snapshot of current staledClientKeys, during deletion new keys may append to staledClientKeys (this is less likely after adding mutex lock before append, still better to keep it safe).

Reply:
What's the expected behavior for new coming keys during deletion? Should we delete them as well or keep them?

Reply:
we should keep them for a while and wait for timer job runs to clean up in later run.